### PR TITLE
WIP Add option to use object store metadata registry

### PR DESCRIPTION
**Pull request by jklegar**
Signed-off-by: Jacob Klegar <jacob@tecton.ai>

**What this PR does / why we need it**: Add support for an object store registry to the SDK. However, note that this should not be used until serving also supports this registry.

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: protos/feast/core/Registry.proto [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import "feast/core/Entity.proto";
+import "feast/core/FeatureTable.proto";
+import "google/protobuf/timestamp.proto";
+
+message Registry {
+    repeated Entity entities = 1;
+    repeated FeatureTable feature_tables = 2;
+
+    string registry_schema_version = 3; // to support migrations
+    string version = 4; // version id, random string generated on each update
```
Comment (woop) [inline protos/feast/core/Registry.proto:33]: What problem does this random string solve, and can it be replaced with a hash of the objects?

File: protos/feast/core/Registry.proto [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+option java_outer_classname = "RegistryProto";
+option go_package = "github.com/feast-dev/feast/sdk/go/protos/feast/core";
+
+import "feast/core/Entity.proto";
+import "feast/core/FeatureTable.proto";
+import "google/protobuf/timestamp.proto";
+
+message Registry {
+    repeated Entity entities = 1;
+    repeated FeatureTable feature_tables = 2;
```
Comment (woop) [inline protos/feast/core/Registry.proto:30]: Is the idea that we will first implement this registry and then come back to add FeatureViews?

File: sdk/python/feast/client.py [hunk 3]
```diff
@@ -394,12 +404,15 @@ def list_projects(self) -> List[str]:
...
         """
 
-        response = self._core_service.ListProjects(
-            ListProjectsRequest(),
-            timeout=self._config.getint(opt.GRPC_CONNECTION_TIMEOUT),
-            metadata=self._get_grpc_metadata(),
-        )  # type: ListProjectsResponse
-        return list(response.projects)
+        if self._use_object_store_registry:
+            raise NotImplementedError("Projects are not implemented for object store registry.")
```
Comment (woop) [inline sdk/python/feast/client.py:408]: How will we accommodate teams that already have projects implemented. Will the migration script remove the project prefix?

File: sdk/python/feast/client.py [hunk 4]
```diff
@@ -146,6 +147,15 @@ def _core_service(self):
             self._core_service_stub = CoreServiceStub(channel)
         return self._core_service_stub
 
+    @property
+    def _use_object_store_registry(self) -> bool:
```
Comment (woop) [inline sdk/python/feast/client.py:151]: I think it's important to have a clean migration path for teams that are adopting the object store registry. If we don't implement the Core reader, then my suggested approach is something like
1. Add object store registry support to SDK (what you are doing)
2. Add a conditional flag like `FALL_BACK_TO_CORE_REGISTRY`. That flag is `True` by default, meaning that  clients will connect to and use the core registry if no registry file is found in the object store during read methods. During create/update operations I think we will have to always create the file, but most of the time the apply methods are controlled by a single team, not by many clients.
3. Teams would migrate by upgrading all client SDKs to the latest version
4. The core registry owners would then shut down their core registry and migrate their configuration to the object store
5. All clients in (3) would then see this new registry file and start using it.
6. The core registry can then be removed entirely.

Comment (woop) [reply sdk/python/feast/client.py:151]: cc @oavdeev

File: sdk/python/feast/registry.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            registry_proto.last_updated.FromDatetime(datetime.utcnow())
+            self._filepath.write_bytes(registry_proto.SerializeToString())
+        return
+
+    def _get_registry(self):
+        registry_proto = RegistryProto()
+        registry_proto.ParseFromString(self._filepath.read_bytes())
+        return registry_proto
+
+    def _update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
```
Comment (woop) [inline sdk/python/feast/registry.py:117]: I need to come back to this code, but I find the classes and methods here surprisingly complex for what is essentially CRUD.

Is there a way that we can simplify the code and make it more readable/explicit by not passing methods around? I'd be quite happy with a simple base class that indicates to implementers what is expected (acts almost like an interface), with an implementation that contains the majority of the logic. 

MLflow for inspiration: https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/abstract_store.py

File: sdk/python/feast/registry.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    @abstractmethod
+    def _get_registry(self):
+        pass
+
+    @abstractmethod
+    def _update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
+        pass
+
+
+class LocalRegistry(Registry):
```
Comment (oavdeev) [inline sdk/python/feast/registry.py:101]: I'd avoid having class hierarchies (except for pure interfaces) as a general rule for better readability. They tend to make things unnecessarily hard to follow when the control flow jumps between the descendant class and the parent. Right now it is fairly simple but as this code will evolve it will likely make things complicated.

One idea would be to separate this into a Registry class (that wouldn't inherit from anything) and maybe RegistryStore(pure ABC), and then LocalRegistryStore would implement get_registry and update_registry. And Registry would use RegistryStore.

File: sdk/python/feast/registry.py (hunk 5 above)
Comment (oavdeev) [reply sdk/python/feast/registry.py:117]: As someone who propsed this interface in the first place; the idea with this was to make it possible for the implementation to use optimistic locking for read-update-write, and to make the interface as minimal as possible. This is a low level class that would have a nicer higher level wrapper (like the Registry above)

File: protos/feast/core/Registry.proto (hunk 1 above)
Comment (oavdeev) [reply protos/feast/core/Registry.proto:33]: It could be a hash, random string is just easier to implement. Properly hashing protos isn't super straightforward

File: sdk/python/feast/registry.py (hunk 5 above)
Comment (woop) [reply sdk/python/feast/registry.py:117]: > This is a low level class that would have a nicer higher level wrapper 

Which class are you referring to here, the `LocalRegistry`?

File: protos/feast/core/Registry.proto (hunk 1 above)
Comment (woop) [reply protos/feast/core/Registry.proto:33]: How will this version be used?

File: sdk/python/feast/registry.py (hunk 5 above)
Comment (oavdeev) [reply sdk/python/feast/registry.py:117]: Yes (i wouldn't have it inherit from Registry tho)

File: protos/feast/core/Registry.proto (hunk 1 above)
Comment (oavdeev) [reply protos/feast/core/Registry.proto:33]: Just debug info for now. For example I would maybe propagate this to tags when creating cloud resources. It is generally nice to have a notion of a "version" for the registry state

File: sdk/python/feast/registry.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        registry_proto = RegistryProto()
+        registry_proto.ParseFromString(self._filepath.read_bytes())
+        return registry_proto
+
+    def _update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
+        registry_proto = self._get_registry()
+        registry_proto = updater(registry_proto)
+        registry_proto.version = str(uuid.uuid4())
+        registry_proto.last_updated.FromDatetime(datetime.utcnow())
+        self._filepath.write_bytes(registry_proto.SerializeToString())
```
Comment (woop) [inline sdk/python/feast/registry.py:122]: Is it a bit weird to use `_get_registry()` which does reading from a file path, but then inside of the `_update_registry()` we have direct access to the file path? Why do we even need _get_registry then?

File: sdk/python/feast/registry.py (hunk 6 above)
Comment (woop) [reply sdk/python/feast/registry.py:101]: @jklegar can we start with this proposal ^?

File: protos/feast/core/Registry.proto (hunk 1 above)
Comment (woop) [reply protos/feast/core/Registry.proto:33]: Ok, so it's almost closer to an id?

Comment (oavdeev) [reply protos/feast/core/Registry.proto:33]: I see, maybe `version_id` would be a clearer name then?

File: sdk/python/feast/client.py (hunk 4 above)
Comment (jklegar) [reply sdk/python/feast/client.py:152]: can we do something simpler like:
1. Team that owns Feast installation creates object store by using the (soon-coming) migration script
2. All clients upgraded to use the object store; either no applies are allowed during this period or any applies done to the core registry must be reflected manually to the object store registry
3. Team removes core registry

File: protos/feast/core/Registry.proto [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    string registry_schema_version = 3; // to support migrations; incremented when schema is changed
+    string version_id = 4; // version id, random string generated on each update of the data; now used only for debugging purposes
```
Comment (woop) [inline protos/feast/core/Registry.proto:33]: Are we actually using it today?

File: sdk/python/feast/client.py (hunk 4 above)
Comment (woop) [reply sdk/python/feast/client.py:152]: It's kind of a hairy upgrade path. What if one team (a Feast user, not owners of system) says that they have a critical production system and they can't upgrade their client for at least 2-3 weeks? That means you need to take Feast Core down for 2-3 weeks while you wait for that team to upgrade. It's not always possible to get everyone to upgrade at the same point in time, so naturally different teams will upgrade at different points in time. The easiest is to provide some kind of middle ground.

File: sdk/python/feast/registry.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+from google.auth.exceptions import DefaultCredentialsError
+
+from feast.core.Registry_pb2 import Registry as RegistryProto
+from feast.entity import Entity
+from feast.feature_table import FeatureTable
+
+REGISTRY_SCHEMA_VERSION = "1"
+
+
+class Registry:
```
Comment (woop) [inline sdk/python/feast/registry.py:32]: Would you mind please adding [Python docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) to classes and methods? You can use your judgement on how much/little detail you want to provide, since some of the methods are documented elsewhere.

File: sdk/python/feast/registry.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        return
+
+    def _write_registry(self, registry_proto: RegistryProto):
+        registry_proto.version_id = str(uuid.uuid4())
+        registry_proto.last_updated.FromDatetime(datetime.utcnow())
+        self._filepath.write_bytes(registry_proto.SerializeToString())
+        return
+
+
+class GCPRegistryStore(RegistryStore):
```
Comment (woop) [inline sdk/python/feast/registry.py:175]: Should this be GCS Registry Store?

File: sdk/python/feast/registry.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        try:
+            from google.cloud import storage
+            from google.cloud.exceptions import NotFound
+        except ImportError:
+            raise ImportError(
+                "Install package google-cloud-storage==1.20.* for gcs support"
+                "run ```pip install google-cloud-storage==1.20.*```"
+            )
+        try:
+            self.gcs_client = storage.Client(project="kf-feast")
```
Comment (woop) [inline sdk/python/feast/registry.py:186]: Is this `kf-feast` a mistake?

File: sdk/python/feast/registry.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        self._blob = self._uri.path.lstrip("/")
+        try:
+            bucket = self.gcs_client.get_bucket(self._bucket)
+        except NotFound:
+            bucket = self.gcs_client.create_bucket(self._bucket)
+            print(bucket)
+        if not storage.Blob(bucket=bucket, name=self._blob).exists(self.gcs_client):
+            registry_proto = RegistryProto()
+            registry_proto.registry_schema_version = REGISTRY_SCHEMA_VERSION
+            self._write_registry(registry_proto)
```
Comment (woop) [inline sdk/python/feast/registry.py:200]: I think it's a bit of a trade off on when we would try and read/write to the registry. My intuition is to do this only when the user runs the `apply()` commands. Does writing the proto here mean that when the user creates an instance of `Client`, we will start writing to the object store?

File: sdk/python/tests/test_client.py [hunk 13]
```diff
@@ -448,6 +515,82 @@ def test_apply_feature_table_success(self, test_client):
             and feature_tables[0].entities[0] == "fs1-my-entity-1"
         )
 
+    @pytest.mark.integration
+    @pytest.mark.parametrize(
+        "test_client", [lazy_fixture("client_with_object_registry_gcs")],
+    )
+    def test_apply_feature_table_integration(self, test_client):
+
+        test_client.set_project("project1")
```
Comment (woop) [inline sdk/python/tests/test_client.py:524]: Is `project1` supposed to be used somewhere?

File: protos/feast/core/Registry.proto (hunk 8 above)
Comment (jklegar) [reply protos/feast/core/Registry.proto:33]: defer to @oavdeev here but I think the idea is we can see it when we look at the stored proto if we need to debug something

File: sdk/python/tests/test_client.py (hunk 13 above)
Comment (jklegar) [reply sdk/python/tests/test_client.py:524]: this is copied from `test_apply_feature_table_success`; I checked and it's not used/needed in either

File: sdk/python/feast/registry.py [hunk 14]
```diff
@@ -30,10 +30,21 @@
 
 
 class Registry:
+    """
+    Registry: Used for interfacing with the object store registry.
```
Comment (woop) [inline sdk/python/feast/registry.py:34]: Sorry, another nitpick. `Foo is an interface for interacting with a foo` is not really descriptive enough. And calling it an interface is also a little bit misleading.

How about `A registry allows for the management and persistence of feature definitions and related metadata`?

File: sdk/python/feast/registry.py [hunk 15]
```diff
@@ -136,27 +209,44 @@ def updater(registry_proto: RegistryProto):
 
 
 class RegistryStore(ABC):
+    """
+    RegistryStore: specific implementations of the object store registry for local file system and GCS.
```
Comment (woop) [inline sdk/python/feast/registry.py:213]: Is this a typo? If this is an ABC is it really a specific implementation?

Comment (woop) [reply sdk/python/feast/registry.py:213]: Also, it seems like the Registry and Registry Store comments are overlapping.
